packages/next/src/server/web/internal-edge-wait-until.ts

```diff
@@ -1,59 +1,30 @@
 // An internal module to expose the "waitUntil" API to Edge SSR and Edge Route Handler functions.
 // This is highly experimental and subject to change.
 
-// We still need a global key to bypass Webpack's layering of modules.
-const GLOBAL_KEY = Symbol.for('__next_internal_waitUntil__')
+const tasks: (() => Promise<any>)[] = []
 
-const state: {
-  waitUntilCounter: number
-  waitUntilResolve: () => void
-  waitUntilPromise: Promise<void> | null
-} =
-  // @ts-ignore
-  globalThis[GLOBAL_KEY] ||
-  // @ts-ignore
-  (globalThis[GLOBAL_KEY] = {
-    waitUntilCounter: 0,
-    waitUntilResolve: undefined,
-    waitUntilPromise: null,
-  })
+async function resolveTasksSequentially(): Promise<void> {
+  let result = Promise.resolve()
 
-// No matter how many concurrent requests are being handled, we want to make sure
-// that the final promise is only resolved once all of the waitUntil promises have
-// settled.
-function resolveOnePromise() {
-  state.waitUntilCounter--
-  if (state.waitUntilCounter === 0) {
-    state.waitUntilResolve()
-    state.waitUntilPromise = null
+  function handleNext() {
+    if (tasks.length) {
+      const task = tasks.shift()
+      result = result.then(() => {
+        handleNext()
+        return task?.()
+      })
+    }
   }
+
+  handleNext()
+
+  return result
 }
 
 export function internal_getCurrentFunctionWaitUntil() {
-  return state.waitUntilPromise
+  return resolveTasksSequentially()
 }
 
-export function internal_runWithWaitUntil<T>(fn: () => T): T {
-  const result = fn()
-  if (
-    result &&
-    typeof result === 'object' &&
-    'then' in result &&
-    'finally' in result &&
-    typeof result.then === 'function' &&
-    typeof result.finally === 'function'
-  ) {
-    if (!state.waitUntilCounter) {
-      // Create the promise for the next batch of waitUntil calls.
-      state.waitUntilPromise = new Promise<void>((resolve) => {
-        state.waitUntilResolve = resolve
-      })
-    }
-    state.waitUntilCounter++
-    return result.finally(() => {
-      resolveOnePromise()
-    })
-  }
-
-  return result
+export function internal_runWithWaitUntil<T>(fn: () => Promise<T>) {
+  tasks.push(fn)
 }
```

Comment:
For visibility: this ensures that those tasks are not started right away - when you call `internal_runWithWaitUntil`, but after the route handler's execution.